docs/flip/2434-general-metadata.md

```diff
@@ -0,0 +1,132 @@
+# FLIP: Axis Metadata
+
+
+- Start Date: 2022-08-08
+- FLIP Issue: [#2434](https://github.com/google/flax/issues/2434)
+- FLIP PR: [#2435](https://github.com/google/flax/pull/2435)
+- Status: Proposal
+
+
+## Summary
+
+This FLIP proposes to extend Flax's variable collections with a generic axis metadata API. 
+The core of the API is a abstract base class that is recognized by lifting transformations that can add an axis (vmap, scan).
+Users can extend the base class to keep track of per-axis metadata in a way that works with lifted transformations.
+
+
+## Motivation
+
+Generally, there is no way in Flax to track metadata for variables across lifted transformations. Axis metadata is used to keep track of semantic information about axes into other (Flax independent) APIs.
+For example, optimizers like AdaFactor can be configured on a per-axis level and partitioning APIs in JAX like xmap or pjit require per variable annotations to map effectiently to parallel hardware.
+
+Currently, there is experimental support for partitioning annotations which requires using dedicated wrapper around lifted transforms that change axes (``nn.scan``, ``nn.vmap``) and a special APIs to create variables (``param_with_axes`` and ``variable_with_axes``). 
```

Comment:
> which requires using dedicated wrapper around lifted transforms that change axes

I don't understand this, could you clarify? Which "dedicated" wrapper are you talking about?

Reply:
Clarified